app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataController.kt

```diff
@@ -125,70 +146,73 @@ class MapDataController internal constructor(
 
     fun getMapDataWithGeometry(bbox: BoundingBox): MutableMapDataWithGeometry {
         val time = currentTimeMillis()
-        val elements = elementDB.getAll(bbox)
-        /* performance improvement over geometryDB.getAllEntries(elements): no need to query
-           nodeDB twice (once for the element, another time for the geometry */
-        val elementGeometries = geometryDB.getAllEntries(
-            elements.mapNotNull { if (it !is Node) ElementKey(it.type, it.id) else null }
-        ) + elements.mapNotNull { if (it is Node) it.toElementGeometryEntry() else null }
-        val result = MutableMapDataWithGeometry(elements, elementGeometries)
-        result.boundingBox = bbox
-        Log.i(TAG, "Fetched ${elements.size} elements and geometries in ${currentTimeMillis() - time}ms")
+        val result = cache.getMapDataWithGeometry(bbox)
+        Log.i(TAG, "Fetched ${result.size} elements and geometries in ${currentTimeMillis() - time}ms")
+
         return result
     }
 
-    private fun Node.toElementGeometryEntry() =
-        ElementGeometryEntry(type, id, ElementPointGeometry(position))
-
     data class ElementCounts(val nodes: Int, val ways: Int, val relations: Int)
+    // this is used after downloading one tile with auto-download, so we should always have it cached
     fun getElementCounts(bbox: BoundingBox): ElementCounts {
-        val keys = elementDB.getAllKeys(bbox)
+        val data = getMapDataWithGeometry(bbox)
         return ElementCounts(
-            keys.count { it.type == ElementType.NODE },
-            keys.count { it.type == ElementType.WAY },
-            keys.count { it.type == ElementType.RELATION }
+            data.count { it is Node },
+            data.count { it is Way },
+            data.count { it is Relation }
         )
     }
 
-    fun getNode(id: Long): Node? = nodeDB.get(id)
-    fun getWay(id: Long): Way? = wayDB.get(id)
-    fun getRelation(id: Long): Relation? = relationDB.get(id)
+    fun getNode(id: Long): Node? = cache.getNode(id) ?: nodeDB.get(id)
+    fun getWay(id: Long): Way? = get(ElementType.WAY, id) as? Way
+    fun getRelation(id: Long): Relation? = get(ElementType.RELATION, id) as? Relation
 
-    fun getAll(elementKeys: Collection<ElementKey>): List<Element> = elementDB.getAll(elementKeys)
+    fun getAll(elementKeys: Collection<ElementKey>): List<Element> =
+        cache.getElements(elementKeys, elementDB::getAll)
 
-    fun getNodes(ids: Collection<Long>): List<Node> = nodeDB.getAll(ids)
-    fun getWays(ids: Collection<Long>): List<Way> = wayDB.getAll(ids)
-    fun getRelations(ids: Collection<Long>): List<Relation> = relationDB.getAll(ids)
+    fun getNodes(ids: Collection<Long>): List<Node> {
+        val nodes = cache.getNodes(ids)
+        return if (ids.size == nodes.size) nodes
+        else {
+            val cachedNodeIds = nodes.map { it.id }
+            nodes + nodeDB.getAll(ids.filterNot { it in cachedNodeIds })
+        }
+    }
```

Comment:
This implementation uses internal knowledge about how the cache works. I would find it cleaner if this was just 
`cache.getNodes(ids, nodesDB::getAll)` like for the others.

I know, the cache does not actually hold nodes outside the cached tiles, but this class doesn't need to know of this.